File: gospider/task/scheduler.go
package task

import (
	"time"
	"sync"
	"log"
)

// Task 定时任务
type Task struct {
	Name     string
	Cron     string
	Handler  func()
	running  bool
	stopChan chan struct{}
}

// Scheduler 任务调度器
type Scheduler struct {
	tasks  []*Task
	mutex  sync.RWMutex
	running bool
}

// NewScheduler 创建任务调度器
func NewScheduler() *Scheduler {
	return &Scheduler{
		tasks:  make([]*Task, 0),
		running: false,
	}
}

// AddTask 添加定时任务
func (s *Scheduler) AddTask(name, cron string, handler func()) *Task {
	task := &Task{
		Name:     name,
		Cron:     cron,
		Handler:  handler,
		stopChan: make(chan struct{}),
	}
	
	s.mutex.Lock()
	defer s.mutex.Unlock()
	s.tasks = append(s.tasks, task)
	
	return task
}

// StartTask 启动任务
func (s *Scheduler) StartTask(task *Task) {
	task.running = true
	
	go func() {
		ticker := time.NewTicker(parseCron(task.Cron))
		defer ticker.Stop()
		
		for {
			select {
			case <-ticker.C:
				log.Printf("[%s] Executing task", task.Name)
				task.Handler()
			case <-task.stopChan:
				log.Printf("[%s] Task stopped", task.Name)
				return
			}
		}
	}()
}

// StopTask 停止任务
func (s *Scheduler) StopTask(task *Task) {
	if task.running {
		task.running = false
		close(task.stopChan)
	}
}

// Start 启动所有任务
func (s *Scheduler) Start() {
	s.running = true
	
	s.mutex.RLock()
	defer s.mutex.RUnlock()
	
	for _, task := range s.tasks {
		s.StartTask(task)
	}
}

// Stop 停止所有任务
func (s *Scheduler) Stop() {
	s.running = false
	
	s.mutex.RLock()
	defer s.mutex.RUnlock()
	
	for _, task := range s.tasks {
		s.StopTask(task)
	}
}

// parseCron 解析 Cron 表达式（简化版）
func parseCron(cron string) time.Duration {
	// 支持格式：*/n * * * * (每 n 分钟)
	// 这里简化实现，实际应该使用完整的 cron 解析库
	return time.Minute
}

// TimedTask 延时任务
type TimedTask struct {
	delay time.Duration
	handler func()
}

// ScheduleTask 调度任务
func ScheduleTask(delay time.Duration, handler func()) *TimedTask {
	task := &TimedTask{
		delay: delay,
		handler: handler,
	}
	
	go func() {
		time.Sleep(delay)
		handler()
	}()
	
	return task
}

// CronTask Cron 任务
type CronTask struct {
	interval time.Duration
	handler func()
	stopChan chan struct{}
}

// NewCronTask 创建 Cron 任务
func NewCronTask(interval time.Duration, handler func()) *CronTask {
	return &CronTask{
		interval: interval,
		handler: handler,
		stopChan: make(chan struct{}),
	}
}

// Start 启动 Cron 任务
func (ct *CronTask) Start() {
	go func() {
		ticker := time.NewTicker(ct.interval)
		defer ticker.Stop()
		
		for {
			select {
			case <-ticker.C:
				ct.handler()
			case <-ct.stopChan:
				return
			}
		}
	}()
}

// Stop 停止 Cron 任务
func (ct *CronTask) Stop() {
	close(ct.stopChan)
}

File: pyspider/task/scheduler.py
"""
定时任务模块
支持 Cron 表达式和延时任务
"""

import threading
import time
from typing import Callable, Dict, List, Optional
from datetime import datetime
from dataclasses import dataclass
import re


@dataclass
class Task:
    """任务"""
    name: str
    cron: str
    handler: Callable
    running: bool = False
    stop_event: Optional[threading.Event] = None


class Scheduler:
    """任务调度器"""
    
    def __init__(self):
        self.tasks: List[Task] = []
        self.running = False
        self._lock = threading.Lock()
        self._threads: List[threading.Thread] = []
    
    def add_task(self, name: str, cron: str, handler: Callable) -> Task:
        """添加定时任务
        
        Args:
            name: 任务名称
            cron: Cron 表达式 (*/n * * * *)
            handler: 处理函数
        """
        task = Task(name=name, cron=cron, handler=handler)
        
        with self._lock:
            self.tasks.append(task)
        
        return task
    
    def start_task(self, task: Task) -> None:
        """启动任务"""
        task.running = True
        task.stop_event = threading.Event()
        
        def run():
            interval = self._parse_cron(task.cron)
            
            while task.running and not task.stop_event.is_set():
                try:
                    task.handler()
                except Exception as e:
                    print(f"[{task.name}] Error: {e}")
                
                task.stop_event.wait(interval)
        
        thread = threading.Thread(target=run, daemon=True)
        thread.start()
        self._threads.append(thread)
    
    def stop_task(self, task: Task) -> None:
        """停止任务"""
        task.running = False
        if task.stop_event:
            task.stop_event.set()
    
    def start(self) -> None:
        """启动所有任务"""
        self.running = True
        
        with self._lock:
            for task in self.tasks:
                self.start_task(task)
    
    def stop(self) -> None:
        """停止所有任务"""
        self.running = False
        
        with self._lock:
            for task in self.tasks:
                self.stop_task(task)
        
        # 等待线程结束
        for thread in self._threads:
            thread.join(timeout=1)
    
    def _parse_cron(self, cron: str) -> float:
        """解析 Cron 表达式（简化版）
        
        支持格式：*/n * * * * (每 n 分钟)
        """
        # 匹配 */n 格式
        match = re.match(r'\*/(\d+)\s+\*\s+\*\s+\*\s+\*', cron)
        if match:
            minutes = int(match.group(1))
            return minutes * 60
        
        # 默认每分钟
        return 60
    
    def get_stats(self) -> Dict:
        """获取统计"""
        return {
            "total_tasks": len(self.tasks),
            "running_tasks": sum(1 for t in self.tasks if t.running),
        }


class TimedTask:
    """延时任务"""
    
    def __init__(self, delay: float, handler: Callable):
        self.delay = delay
        self.handler = handler
        self._thread: Optional[threading.Thread] = None
        self._cancelled = False
    
    def start(self) -> None:
        """启动任务"""
        def run():
            time.sleep(self.delay)
            if not self._cancelled:
                self.handler()
        
        self._thread = threading.Thread(target=run, daemon=True)
        self._thread.start()
    
    def cancel(self) -> None:
        """取消任务"""
        self._cancelled = True


def schedule_task(delay: float, handler: Callable) -> TimedTask:
    """调度延时任务"""
    task = TimedTask(delay, handler)
    task.start()
    return task


class CronTask:
    """Cron 任务"""
    
    def __init__(self, interval: float, handler: Callable):
        self.interval = interval
        self.handler = handler
        self._running = False
        self._thread: Optional[threading.Thread] = None
        self._stop_event = threading.Event()
    
    def start(self) -> None:
        """启动任务"""
        self._running = True
        
        def run():
            while self._running and not self._stop_event.is_set():
                try:
                    self.handler()
                except Exception as e:
                    print(f"CronTask Error: {e}")
                
                self._stop_event.wait(self.interval)
        
        self._thread = threading.Thread(target=run, daemon=True)
        self._thread.start()
    
    def stop(self) -> None:
        """停止任务"""
        self._running = False
        self._stop_event.set()
